### ISSUE TITLE
Play next video automatically, next video is in background

### ISSUE THREAD

**Issue by Portisch** [xbmc/xbmc#19153]
### Describe the bug
Here is a clear and concise description of what the problem is:

Since the commit https://github.com/xbmc/xbmc/commit/4f8a656aed05ff6d63aacac0d3a610381a4343c4 the video what is started after one is finished is played in background of GUI.
@djp952 

## Expected Behavior
Here is a clear and concise description of what was expected to happen:

Next video just should start playing without showing the GUI.

## Actual Behavior

When video ends GUI shows up and next video start playing in background.

## Possible Fix

Revert https://github.com/xbmc/xbmc/commit/4f8a656aed05ff6d63aacac0d3a610381a4343c4

### To Reproduce

1. enable play next video for on option minimum
2. start a playback and forward near to the end
3. wait till next video getting started
4. GUI show up and video start playing in background

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [X] Linux
 - [ ] OSX
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name:
 CoreELEC kernel 4.9.119
 - Kodi version:
Last xbmc/master HEAD b69b6ea3d84c6dd1d3425cd5349c8dc635657985

Issue is platform independent.

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by wsnipex** [comment]
confirmed on linux and android (aarch64/shield)

**Comment by popcornmix** [comment]
I've seen this on Pi4 (gbm) when using SendToKodi to add a youtube video using "add to playlist" option with nightly build. It didn't happen a month or so back.

**Comment by djp952** [comment]
I'll see if I can reproduce and propose a solution; https://github.com/xbmc/xbmc/commit/4f8a656aed05ff6d63aacac0d3a610381a4343c4 resets member variables of VideoPlayer when it's being (re)prepared, I don't think rolling that back would be the proper way to fix this.  

Note that I only have Windows available to test builds on, if it's a state issue with VideoPlayer it should happen on all platforms I would think.

**Comment by djp952** [comment]
I can reproduce and confirm that not resetting m_HasVideo here will solve it, although m_HasVideo is being set to true a bit later on, it could be a simple fix.  Give me until after work :)

**Comment by djp952** [comment]
I don't think the problem is the change in and of itself, I think it's more that it exposed a race condition that has this ill effect when a playlist is involved.  I added some logging to illustrate this.  The Main Thread (9952 here) is interrogating m_hasVideo after it's been reset to false, but before VideoPlayer has a chance to set it again:

```
2021-02-01 17:57:48.214 T:4284     INFO <general>: CVideoPlayer::Prepare() - reset m_HasVideo
2021-02-01 17:57:48.214 T:9952     INFO <general>: CVideoPlayer::HasVideo()
2021-02-01 17:57:48.214 T:4284     INFO <general>: Creating InputStream
2021-02-01 17:57:48.214 T:9952     INFO <general>: CVideoPlayer::HasVideo()
2021-02-01 17:57:48.231 T:4284     INFO <general>: Creating Demuxer
2021-02-01 17:57:48.258 T:4284     INFO <general>: Opening stream: 0 source: 256
2021-02-01 17:57:48.258 T:4284     INFO <general>: Creating video codec with codec id: 27
2021-02-01 17:57:48.258 T:4284     INFO <general>: CDVDVideoCodecFFmpeg::Open() Using codec: H.264 / AVC / MPEG-4 AVC / MPEG-4 part 10
2021-02-01 17:57:48.258 T:4284     INFO <general>: Creating video thread
2021-02-01 17:57:48.260 T:12332    INFO <general>: running thread: video_thread
2021-02-01 17:57:48.260 T:4284     INFO <general>: CVideoPlayer::OpenVideoStream() - set m_HasVideo
```

Rolling back the change just allows VideoPlayer to report false information.  Until Prepare() is called, it doesn't know if the file has a video (or audio) stream, it just continues to return the state of the previous stream.  When playing a video playlist, this is perfectly fine, m_HasVideo will stay true and then just be reset to true again.  However, I'd bet if you threw a file with no video stream into that playlist the original problem will appear, you would get a black screen (or on Windows a still frame), and the UI wouldn't switch to audio only mode.

VideoPlayer can also reset the state of m_HasVideo (and m_HasAudio) if a stream change packet is detected, that is also not protected from any race with HasVideo().

One suggestion might be to lock m_HasVideo and m_HasAudio, my concern with this would be that during an operation like Prepare() that lock could be held quite a while (looks like ~50ms on my machine) and it might cause problems with the expectation that HasVideo() will return immediately.

Another suggestion would be to find why a playlist behaves differently than say, a PVR channel change, and see if something can be inserted there to allow VideoPlayer to initialize all the way before interrogating it.  That wouldn't necessarily prevent any race condition on a stream change packet, but it doesn't appear anyone has complained about that either :)

I'd like to know what @FernetMenta thinks about this if he has time.

Happy to add something else or roll this back if that's the way everyone wants to go.  I do honestly think it's just uncovered a different problem and isn't necessarily to blame.

**Comment by FernetMenta** [comment]
Over the past years VP has adapted to many new scenarios but the surrounding CApp still thinks it controls the old dvd player. The content VP plays can change at any time and the UI should adapt to it. m_HasVideo / m_HasAudio are relics, redundant, and shouldn't be used anymore. Methods to be used are in the content section of VP: GetVideoStreamCount / GetAudioStreamCount
When there are changes to video/audio/subs VP updates content section and signals an AV change in CDataCacheCore. DataCacheCore was built to serve guiInfoManager and is not ideal to be used by CApp. I would call the OnAvChange callback function after VP has updated the content. Then CApp can adapt on GUI_MSG_PLAYBACK_AVCHANGE. CApp needs to track player's state and act accordingly (how the event GUI_MSG_PLAYBACK_AVCHANGE should be handled depends on the state if CApp has already received the AVStarted event).

**Comment by wsnipex** [comment]
@FernetMenta given that we nearly final with v19, would you recommend this as the correct fix for v19 or would it require larger refactoring that should be postponed to v20?

**Comment by FernetMenta** [comment]
I recommend it as the correct fix but I won't give recommendations on what version you want to schedule required changes.

**Comment by djp952** [comment]
Do you want to try something like a lock for v19? I could work that up this evening, if it works I think the bulk of comments/fixes to that would be scope related and shouldn't take a long time overall.

My opinion, for what it's worth, is that significantly more people would be negatively impacted by the playlist problem it's causing than would benefit from the original change.  If there is no relatively simple solution available for v19 I completely agree with undoing the change and waiting for v20 to fix it all properly.

**Comment by djp952** [comment]
My proposal for using a lock to solve this without undoing the commit is here, if anyone wants to look at it: https://github.com/djp952/xbmc/commit/f69bd072c0207a9a437bb1bfa198602d0fab7bba.

I found that it solves the problem without regressing the original intention, but I have a couple concerns:

- The lock scope is a pessimistic in places and needs to be reentrant (which it is).  I think it could be tightened by changing the bulk of the functions/member variables in VideoPlayer from "protected" to "private", I had no problems inserting a **private:** label before this location, for example:

https://github.com/xbmc/xbmc/blob/0c9370c02202b2020ffb6d05842c1c4c6872194a/xbmc/cores/VideoPlayer/VideoPlayer.h#L359-L360

- This lock may be held longer than can be anticipated/expected by callers to HasAudio() or HasVideo().  I found no instances where that was problematic in any way, but clearly I am not capable of anticipating all potential impacts of such a change.

My argument for this, or some variant of this, is that it does seem to solve the reported problem and should be safe enough to be considered for v19 without undoing the original PR.  My argument against this is that these flags/functions should be eliminated rather than tweaked/fixed.

Whichever way is deemed best for v19, just let me know and I'll take the necessary actions to get this off of the Issue list.

**Comment by wsnipex** [comment]
I'd vote for reverting that offending patch, fix it correctly for v20, then backport to v19 point release when it was sufficiently tested in master.

**Comment by Portisch** [comment]
Just my 2¢:
If a lock will solve it I would prefer to fix it properly.
We were running in past v18 already in the same issue when a video plays and another one get started without stopping the old one before. Then the next video decoder is loaded and short after the old get unloaded -> lock issue.

The Amlogic hardware decoder is capable to decode up to 9 streams in parallel. It was needed to limit the system hardware decoder to 1 stream as otherwise Kodi opened decoder[0] and then decoder[1]. So the stream is on decoder index 1 but Kodi is only able to handle one stream on index 0. The result was black picture as the Kodi used the wrong decoder.

So like to see this as a lock issue it needs to be fixed correctly. Like for PiP this fix will be needed anyway.

**Comment by FernetMenta** [comment]
> djp952@f69bd07

The lock is the most horrible approach one can think of in this case. It might look less intrusive but it is not. You lock long running operations like opening some stream/file on a remote location. This can take longer than 20s or even more.

**Comment by djp952** [comment]
Will submit PR to revert the change
